Chunking RFC (#2)
o1-labs/rfcs: 0002-chunking.md

> ## Summary

Reviewer: ```suggestion
## Chunking summary
```

Reviewer: I suggest we change "Summary" to the gist of the RFC so we can identify the subject matter at a glance, otherwise, all of our RFCs will be titled "Summary" (not very useful)

Author: @bkase wdyt for the template? Should we propose a structure that customises the summary for each?

Reviewer 2: We could add a top-level `#` heading with the title of the feature maybe? But then keep `Summary` because it's a standard. What do you think @barriebyron ?

Reviewer: better, yes. I updated the template, too

## Motivation
[motivation]: #motivation

The motivation, assessment criteria, and exit criteria can be found in the [PRD](https://www.notion.so/minaprotocol/Chunked-Commitments-PRD-890c836529c545f082a3eeecd3d4f510).

This proposal raises the maximum 'circuit size' limit for kimchi proofs from `2^16` to `2^32`. Out-of-scope is further work which would remove this limit, since this limit will not currently be reachable without other improvements to memory usage and performance.

This proposal optimizes for compatibility with the kimchi and pickles [versions proposed for deployment in the Berkeley hard-fork](https://github.com/MinaProtocol/MIPs/blob/main/MIPS/mip-kimchi.md), as well as the [proposed zkApps feature](https://github.com/MinaProtocol/MIPs/blob/main/MIPS/mip-zkapps.md), at the expense of some potential loss of performance when the feature is in use.  Timing measurements can be found in the PR description for the [prototype implementation](https://github.com/o1-labs/proof-systems/pull/1033).

> ## Detailed design

Reviewer 2: Does this impact anything other than Pickles/Kimchi and SnarkyJS (and the bindings)? Probably not, but just double checking. Maybe worth calling out explicitly? I tried to do that in the RFC I just drafted.

[detailed-design]: #detailed-design

> ### Theory overview

Reviewer 2: This is amazing content, but maybe it's better served as a chapter of the Mina book and linked to from this RFC, so it's more accessible? What do you think?

This is of course not clear from the current template, so we can adjust after we settle this.

```
which is equivalent to committing to the 4 'chunked' polynomials
```text
f_0(x) = a_0 x^0 + a_1 x^1 + ... + a_{n-1} x^{n-1}
f_1(x) = a_n x^0 + a_{n+1} x^1 + ... + a_{2n-1} x^{n-1}
f_2(x) = a_{2n} x^0 + a_{2n+1} x^1 + ... + a_{3n-1} x^{n-1}
f_3(x) = a_{3n} x^0 + a_{3n+1} x^1 + ... + a_{4n-1} x^{n-1}
```
that we can recompose to form `f`
```text
f(x) = f_0(x) + x^n * f_1(x) + x^{2n} * f_2(x) + x^{3n} * f_3(x)
```

When we check the constraint polynomial at a ['random'](https://en.wikipedia.org/wiki/Fiat%E2%80%93Shamir_heuristic) point `z` later in the proof, we can build a commitment that matches that point by computing
```text
C = C_0 + C_1^{z^n} + C_2^{z^{2n}} + C_3^{z^{3n}}
```
while also checking that the prover's claimed evaluations are consistent with the commitments:
```text
[( f_0(z), C_0 ), ( f_1(z), C_1 ), ( f_2(z), C_2 ), ( f_3(z), C_3 )]
```
Finally, we can compute the value `f(z)` by proxy:
```text
f(z) = f_0(z) + z^n f_1(z) + z^{2n} f_2(z) + z^{3n} f_3(z)
```

Thus, this chunking approach provides all of the features that we need to support the kimchi proof system over a larger domain size than the current Mina limit.

### Retaining zero-knowledge

> Once we have `c > 1` chunks, every column will come with `2 * c` evaluations, 1 for each chunk at each of `z` and `z * w`. In order to retain zero-knowledge, we need to ensure that we randomize more columns than there are evaluations. Thus, we must randomize at least `2 * c + 1` values at the end of each column. The helpers for adding zero-knowledge to columns should be updated to take the number of zk rows that are required as an argument.

Reviewer: ```suggestion
After we have `c > 1` chunks, every column will come with `2 * c` evaluations, 1 for each chunk at each of `z` and `z * w`. In order to retain zero-knowledge, we need to ensure that we randomize more columns than there are evaluations. Thus, we must randomize at least `2 * c + 1` values at the end of each column. The helpers for adding zero-knowledge to columns must be updated to take the number of zk rows that are required as an argument.
```

Reviewer: if we mean "must" let's be sure to use must, otherwise I am reading "should" as a suggestion

In total, this modified permutation argument achieves the same zero knowledge

#### Number of zk rows

The permutation argument interacts with the `c` chunks in parallel, so it is
possible to cross-correlate between them to compromise zero knowledge. We know
that there is some `c >= 1` such that `zk_rows = 2*c + k` from the above. Thus,
attempting to find the evaluation at a new point, we find that:
* the evaluation of every witness column in the permutation contains `k` unknowns;
* the evaluations of the permutation argument aggregation has `k-1` unknowns;
* the permutation argument applies on all but `zk_rows - 3` rows;
* and thus we form the equation `zk_rows - 3 < 7 * k + (k - 1)` to ensure that we
  can construct fewer equations than we have unknowns.

This simplifies to `k > (2 * c - 2) / 7`, giving `zk_rows > (16 * c - 2) / 7`.
We can derive `c` from the `max_poly_size` supported by the URS, and thus we find
`zk_rows` and `domain_size` satisfying the fixpoint

```text
zk_rows = (16 * (domain_size / max_poly_size) + 5) / 7
domain_size = circuit_size + zk_rows
```

> ### Technical approach in kimchi

Reviewer 2: This part I think is definitely perfect to remain in the RFC

> Support for chunked commitments already exists [in kimchi's supporting polynomial commitment library](https://github.com/o1-labs/proof-systems/blob/ca0b9e53d15e3ab5db6fb8c20617979cbf74215e/poly-commitment/src/srs.rs#L89). This RFC proposes to extend that support through to the kimchi prover, where an attempt to use the feature currently [results in an error](https://github.com/o1-labs/proof-systems/blob/ca0b9e53d15e3ab5db6fb8c20617979cbf74215e/kimchi/src/prover.rs#L170). This RFC also proposes to add support for chunked commitments and evaluations to pickles, where there is currently no support.

Reviewer: I feel like we "buried the lede" here, this detail would be very helpful in RFC summary section

The verifier logic for [the public input commitment](https://github.com/o1-labs/proof-systems/blob/ca0b9e53d15e3ab5db6fb8c20617979cbf74215e/kimchi/src/verifier.rs#L592) will need to use the chunked lagrange basis, to be consistent with how we will generate them on the prover side.

To avoid the `O(c n log(n))` cost of computing the chunked evaluation for the public input polynomial in the verifier, the prover should pass this evaluation as part of the proof. In order to avoid breaking compatibility with zkApps, this 'public input' field should be optional, as proofs submitted in zkApp commands will not have this field available. This value is already considered part of the transcript for the purposes of pickles, so there is no soundness risk of using a value provided by a malicious prover.

> Once all of the above are complete, the error checks in kimchi's prover and verifier can be removed, and chunking can be considered enabled in kimchi. This should be accompanied by a test that exercises this behavior.

Reviewer: ```suggestion
After all of the above changes are complete, the error checks in kimchi's prover and verifier can be removed, and chunking can be considered enabled in kimchi. This should be accompanied by a test that exercises this behavior.
```

### Technical approach in OCaml-to-rust bindings

The polynomial commitments and evaluations contained in the proofs and the verifier keys will need to be represented as `array`s instead of single values. Several of the OCaml type conversion helpers assume that these only ever contain 1 element. The [autogenerated OCaml types](https://github.com/MinaProtocol/mina/blob/c71d0f8852774b718a8f0dd5af37520985e57813/src/lib/crypto/kimchi_bindings/stubs/kimchi_types.ml) should be used to validate this assumption, and to guide the necessary changes if useful.

The additional public input evaluation must not appear in the types already serialized as part of zkApps, including the list of evaluations, where they most naturally would fit. They should be passed in a separate structure, and should be optional as on the rust side.

### Technical approach in pickles

> The pickles code should be updated to reflect the changes to the bindings, which will propagate `array`s into several types. The logic for handling these will then need to be updated to handle these cases, in particular the [`ft_comm` calculation](https://github.com/MinaProtocol/mina/blob/c71d0f8852774b718a8f0dd5af37520985e57813/src/lib/pickles/common.ml#L231) will need to be updated to match [the kimchi logic](https://github.com/o1-labs/proof-systems/blob/ca0b9e53d15e3ab5db6fb8c20617979cbf74215e/kimchi/src/verifier.rs#L705), absorbing the evaluations in the [step verifier](https://github.com/MinaProtocol/mina/blob/c71d0f8852774b718a8f0dd5af37520985e57813/src/lib/pickles/step_verifier.ml#L928) and the [wrap verifier](https://github.com/MinaProtocol/mina/blob/c71d0f8852774b718a8f0dd5af37520985e57813/src/lib/pickles/wrap_verifier.ml#L871), and the `check_bulletproof` logic for [step](https://github.com/MinaProtocol/mina/blob/c71d0f8852774b718a8f0dd5af37520985e57813/src/lib/pickles/step_verifier.ml#L222) and [wrap](https://github.com/MinaProtocol/mina/blob/c71d0f8852774b718a8f0dd5af37520985e57813/src/lib/pickles/wrap_verifier.ml#L388).

Reviewer: ```suggestion
The pickles code must be updated to reflect the changes to the bindings, which will propagate `array`s into several types. The logic for handling these will then need to be updated to handle these cases, in particular the [`ft_comm` calculation](https://github.com/MinaProtocol/mina/blob/c71d0f8852774b718a8f0dd5af37520985e57813/src/lib/pickles/common.ml#L231) will need to be updated to match [the kimchi logic](https://github.com/o1-labs/proof-systems/blob/ca0b9e53d15e3ab5db6fb8c20617979cbf74215e/kimchi/src/verifier.rs#L705), absorbing the evaluations in the [step verifier](https://github.com/MinaProtocol/mina/blob/c71d0f8852774b718a8f0dd5af37520985e57813/src/lib/pickles/step_verifier.ml#L928) and the [wrap verifier](https://github.com/MinaProtocol/mina/blob/c71d0f8852774b718a8f0dd5af37520985e57813/src/lib/pickles/wrap_verifier.ml#L871), and the `check_bulletproof` logic for [step](https://github.com/MinaProtocol/mina/blob/c71d0f8852774b718a8f0dd5af37520985e57813/src/lib/pickles/step_verifier.ml#L222) and [wrap](https://github.com/MinaProtocol/mina/blob/c71d0f8852774b718a8f0dd5af37520985e57813/src/lib/pickles/wrap_verifier.ml#L388).
```

In order to represent chunked proofs in the pickles circuits, it will also be necessary to update the `Typ.t`s for proofs to allow a flexible number of chunks for both step and wrap proofs. The [`Commitment_lengths`](https://github.com/MinaProtocol/mina/blob/c71d0f8852774b718a8f0dd5af37520985e57813/src/lib/pickles/commitment_lengths.ml) and [`Evaluation_lengths`](https://github.com/MinaProtocol/mina/blob/c71d0f8852774b718a8f0dd5af37520985e57813/src/lib/pickles/evaluation_lengths.ml) constants can be modified to cause compiler warnings in most of the relevant locations. The representation of unchunked proofs must not change, including, but not limited to, reordering of any fields. The lengths passed to these `Typ.t`s must be inferrable from the 'tag' that pickles associates with each collection of inductive rules; as such, they should be stored in the `Types_map` system along with all similar metadata.

> Due to the requirement for proof layout homogeneity, the `Pickles.compile` interface should take parameters for the `step_num_chunks` and `wrap_num_chunks` at the toplevel. These arguments may be optional labelled arguments, since they will not be regularly used in the protocol circuits. These arguments will have to be passed explicitly by the user / caller when needed, since it will be prohibitively expensive to try all of the various options programatically on every compile.

Reviewer: ```suggestion
Due to the requirement for proof layout homogeneity, the `Pickles.compile` interface must take parameters for the `step_num_chunks` and `wrap_num_chunks` at the top level. These arguments may be optional labelled arguments, since they will not be regularly used in the protocol circuits. These arguments will have to be passed explicitly by the user / caller when needed, since it will be prohibitively expensive to try all of the various options programmatically on every compile.
```

> The `Pickles.compile` function should be modified to check for the number of chunks in each inductive rule's previous rule tags. When these are not consistent between inductive rules, the code must throw an error in order to achieve layout homogeneity.

Reviewer: ```suggestion
The `Pickles.compile` function must be modified to check for the number of chunks in each inductive rule's previous rule tags. When these numbers are not consistent between inductive rules, the code must throw an error in order to achieve layout homogeneity.
```

Reviewer: @mrmr1993  I love this first PRD, congratulations! What do we mean by "these" (these tags? or these numbers?)

Author: `These numbers` is correct. The pickles recursion layer needs everything to have the 'same shape', but the 'number of chunks' result in different layouts for the proofs that it consumes. Is there a way to word this that would make that detail clearer?

### Prototype PRs

* [Fix chunking for proofs](https://github.com/o1-labs/proof-systems/pull/1032)
* [Update bindings for chunking fix](https://github.com/MinaProtocol/mina/pull/13285)
* [Enable chunked proving](https://github.com/o1-labs/proof-systems/pull/1033)
* [Add zero-knowledge for chunked proofs](https://github.com/o1-labs/proof-systems/pull/1045)
* [Update bindings to enable chunking](https://github.com/MinaProtocol/mina/pull/13286)

> ## Test Plan and Functional Requirements

Reviewer 2: This section needs a little more detail I think -- unit testing what specifically at each layer? Which parts of the system? And then we should sketch out the SnarkyJS test also in a bit more detail, but maybe that entails figuring out the SnarkyJS details first: I suppose something with ZkProgram though?

## Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

A previous alternative was to increasing the circuit size limit by generating a larger (`2^17`) URS. This was rejected after the slow-down for verifying kimchi proofs on Berkeley was deemed unacceptable.

Since this feature was already partly implemented, it is significantly lower-cost than other potential proof-system changes.

None of the other alternatives were able to increase the circuit size limit without either breaking compatibility with zkApps or dramatically increasing complexity and proving time. In particular, switching the polynomial commitments scheme, and using a different proof system were considered and rejected by the crypto team.

Failure to complete this feature will result in less usability for zkApps, which will have a negative effect on both the Berkeley hard fork and SnarkyJS.

## Prior art
[prior-art]: #prior-art

The existing broken chunking implementation.

## Unresolved questions
[unresolved-questions]: #unresolved-questions

* What parts of the design do you expect to resolve through the RFC process before this gets merged?

>   - SnarkyJS design

Reviewer 2: What exactly do we need to expose to the user? The number of chunks? Can't we compute that at circuit compile time? This is important for the SnarkyJS API design
